### ISSUE TITLE
t.i.task.Cooperator should cancel scheduled call if last task is removed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5223]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5223|
|Type|enhancement|
|Created|2011-08-07 01:11:22Z|
|Branch|https://github.com/twisted/twisted/tree/cooperator-cleanup-5223|

When writing tests it is good to have no reactor state left behind. However, Cooperator may leave a DelayedCall running even if the last task in the Cooperator has been removed. It would be better to cancel the scheduled tick if there is no more work for the Cooperator to do.

I encountered this in [#7153](https://github.com/twisted/twisted/issues/7153).

<details><summary>Searchable metadata</summary>

```
trac-id__5223 5223
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_cooperator_cleanup_5223 branches/cooperator-cleanup-5223
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1312679482000000 1312679482000000
changetime__1313279867000000 1313279867000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [32428]) Branching to 'cooperator-cleanup-5223'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Looks good.  If you get a clean build on buildbot, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [32450]) Merge cooperator-cleanup-5223

Author: itamar
Reviewer: exarkun
Fixes: [#5223](https://github.com/twisted/twisted/issues/5223)

In `twisted.internet.task.Cooperator`, cancel the pending scheduling call
as soon as the last task is stopped.  This makes a `Cooperator` more
reactor/testing friendly, by leaving no remaining event sources once it
has no more work to do.
